### PR TITLE
Geometry optimization with fixed atoms

### PR DESCRIPTION
**Pull request by mahrossi**
This is an implementation that makes a (much) better job at fixing atoms for geometry optimizations than what was there before. L-BFGS is not supported but the code stops if trying to do that at the moment. Needs some review to make sure the code is ok, but I have looked over it and it looks good.

Tests and examples are in examples/fhi-aims/h2o-geop-fixatoms (also includes an explicit comparison to ASE -- we can remove it).

### CODE REVIEW

**Comment by mahrossi** [comment]
@litman90 can you review this too?

**Comment by ceriottm** [comment]
Do you mean that l-bfgs is out, or that l-bfgs works but will stop if there
are fixed atoms?

**Comment by fidanyan** [comment]
L-BFGS works as before if there are no fixed atoms, otherwise in throws a warning and stops.

**Comment by litman90** [comment]
Hi,
I have tested trmbfgs fixing 1 atom and it **doesn't** work as expected.
The main problem is that the Hessian is not being updated which arises from the way the masking procedure is performed.

**Comment by fidanyan** [comment]
Hello,
All algorithms work now, the only point to pay attention is that I changed default `ls_options["tolerance"]` from 1 to 1e-2, otherwise L-BFGS got stuck in `min_approx()`, thinking that the position is already converged. I think it's unrelated to fixatoms, even though I could not reproduce this behaviour in other cases. Any comments on `ls_options` are appreciated, also it is not necessary change since it can be provided in input file, so this part can be rejected without heavy consequences.

**Comment by mahrossi** [comment]
I think you need even tighter criteria to converge there in many cases
(1e-4 or so). We had already discussed this long ago and I fixed this point
also in another branch already.

**Comment by mahrossi** [comment]
@fidanyan what is the status here? have you thoroughly tested that _all_ algorithms work with fixed atoms like this?

**Comment by fidanyan** [comment]
Yes, they passed all my tests. examples/lammps/h2o-geop-fixatoms/README and examples/lammps/h2o-geop-fixatoms/README contain the results.

File: examples/fhi-aims/h2o-geop-fixatoms/ase_comparison/log.ase [hunk 1]
```diff
@@ -0,0 +1,1011 @@
+Accepting clients on INET port 31415
```
Comment (mahrossi) [inline examples/fhi-aims/h2o-geop-fixatoms/ase_comparison/log.ase]: This file is probably not needed and will just inflate the repository. These type of files should never be committed.

File: examples/fhi-aims/h2o-geop-fixatoms/ase_comparison/opt.log [hunk 2]
```diff
@@ -0,0 +1,40 @@
+      Step     Time          Energy         fmax
```
Comment (mahrossi) [inline examples/fhi-aims/h2o-geop-fixatoms/ase_comparison/opt.log]: This and the next xyz files should be in a "regression test" type of folder. Not really in the examples folder. Examples are meant to tell people how to run something.

File: examples/lammps/h2o-geop-fixatoms/run-all-tests.sh [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+
+. ~/soft/i-pi-mahrossi/env.sh
```
Comment (mahrossi) [inline examples/lammps/h2o-geop-fixatoms/run-all-tests.sh]: explicit paths in test scripts will obviously make them break for any other user.

**Comment by mahrossi** [comment]
In addition: In the examples folder you can leave one or two examples (only with inputs) about how to run geop with fixed atoms and that should be enough.

**Comment by fidanyan** [comment]
Done: I left only the minimal set in the examples/, and all the result files with an explanations are in tests/benchmarks/*.tgz.

**Comment by fidanyan** [comment]
Moved test cases from tests/benchmarks/ to examples/regtest. `regtest.py` works for lammps, but for FHI-aims it doesn't work for some reason. Maybe later I'll try to sort it out. For now, I just left FHI-aims results as an archive with the explanations inside. It's quite clear.

**Pull request merged**